### PR TITLE
test tail calls execute in constant env-stack

### PR DESCRIPTION
**Pull request by nickchapman-da**
Test that:
- Tail-recursive function definitions execute with bounded environment and continuation stacks.
- also test `FOLDL` and `FOLDR`

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  val unbounded: Option[Int] = None
+
+  "A *non* tail-recursive definition requires an unbounded env-stack, and an unbounded kont-stack" in {
+    val exp = e"F:triangle 100"
+    val expected = SValue.SInt64(5050)
+    // The point of this test is to prove that the bounded-evaluation checking really works.
+    runExpr(exp, envBound = unbounded, kontBound = unbounded) shouldBe expected
+    expectThrows {
+      runExpr(exp, envBound = small, kontBound = unbounded)
+    }
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:66]: Let's be a bit more precise on what gets thrown:
```suggestion
val error = the[RuntimeException] thrownBy runExpr(exp, envBound = small, kontBound = unbounded)
error.toString should include("BoundedArrayList, exceeded bound of 5")
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+       }
+      """
+
+  val small: Option[Int] = Some(5)
+  val unbounded: Option[Int] = None
+
+  "A *non* tail-recursive definition requires an unbounded env-stack, and an unbounded kont-stack" in {
+    val exp = e"F:triangle 100"
+    val expected = SValue.SInt64(5050)
+    // The point of this test is to prove that the bounded-evaluation checking really works.
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:62]: Good thinking!

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+          } else {
+            machine.kontStack.get(0)
+          }
+        machine.kontStack = new BoundedArrayList[Speedy.Kont](bound)
+        machine.kontStack.add(onlyKont)
+    }
+    // run the machine
+    machine.run() match {
+      case SResultFinalValue(v) => v
+      case res => throw new RuntimeException(s"runExpr, unexpected result $res")
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:135]:
```suggestion
      case res => crash(s"runExpr, unexpected result $res")
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  private def typeAndCompile(pkg: Ast.Package): PureCompiledPackages = {
+    val rawPkgs = Map(defaultParserParameters.defaultPackageId -> pkg)
+    Validation.checkPackage(rawPkgs, defaultParserParameters.defaultPackageId, pkg)
+    data.assertRight(
+      PureCompiledPackages(rawPkgs, Compiler.Config.Default.copy(stacktracing = FullStackTrace)))
+  }
+
+  val pkgs = typeAndCompile(pkg)
+
+  private def expectThrows[A](f: => A): Assertion = {
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:99]: This becomes probably obsolete if we use the `the[RuntimeException] thrownBy ...` approach outline above.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  class BoundedArrayList[T](bound: Int) extends util.ArrayList[T](bound) {
+    override def add(x: T): Boolean = {
+      if (size >= bound) {
+        crash(s"BoundedArrayList, exceeded bound of $bound")
+      }
+      super.add(x)
+    }
+  }
+
+  def crash[A](reason: String): A = throw new RuntimeException(reason)
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:148]: Don't you do something like this in a language with subtyping?
```suggestion
  def crash(reason: String): Nothing = throw new RuntimeException(reason)
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-left executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangle_viaFoldLeft 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-right executes with a small env-stack, and a small kont-stack" in {
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:84]: Have you by any chance tried what happens if we hit the `KFoldr1Map/Reduce` case? For instance, if the step function passed to `foldr` is something like `\x -> identity(\y -> ...)`. I'd be very curious to know. 😃

**Review by hurryabit — APPROVED** [review]
Awesome. Thanks a lot!

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala (hunk 5 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:156]: I think either way is acceptable. The parametric form is used in speedy code `SError.scala`

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala (hunk 6 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:93]: I added a test which checks the `KFold1Map/Reduce` works in bounded space.

Comment (hurryabit) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:93]: Nice. Thank you.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+
+         val triangle_viaFoldLeft : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDL @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         val triangle_viaFoldRight : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDR @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         val triangle_viaFoldRight2 : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDR @Int64 @Int64 (\(y: Int64) -> ADD_INT64 y) 0 (F:generate Nil@Int64 x);
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala:48]: That's a nice way to trigger the `KFoldr1Map/Reduce` case. 👍

**Pull request merged**